Classes/Extensions/DDFileLogger+Buffering.m

```diff
@@ -16,146 +16,183 @@
+static NSUInteger DDGetDefaultBufferSizeBytes() {
+    static NSUInteger defaultBufferSize = 0;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        defaultBufferSize = p_DDGetDefaultBufferSizeBytesMax(NO);
+    });
+    return defaultBufferSize;
 }
 
-- (instancetype)initWithInstance:(FileLogger)instance;
+@interface DDBufferedProxy : NSProxy
+
+@property (nonatomic) DDFileLogger *fileLogger;
+@property (nonatomic) NSOutputStream *buffer;
 
-@property (strong, nonatomic, readwrite) FileLogger instance;
+@property (nonatomic) NSUInteger maxBufferSizeBytes;
+@property (nonatomic) NSUInteger currentBufferSizeBytes;
 
 @end
 
-@interface DDBufferedProxy (StreamManipulation)
+@implementation DDBufferedProxy
 
-- (void)flushBuffer;
-- (void)dumpBufferToDisk;
-- (void)appendToBuffer:(NSData *)data;
-- (BOOL)isBufferFull;
+- (instancetype)initWithFileLogger:(DDFileLogger *)fileLogger {
+    _fileLogger = fileLogger;
+    _maxBufferSizeBytes = DDGetDefaultBufferSizeBytes();
+    [self flushBuffer];
 
-@end
+    return self;
+}
+
+- (void)dealloc {
+    dispatch_block_t block = ^{
+        [self lt_sendBufferedDataToFileLogger];
+        self.fileLogger = nil;
+    };
 
-@implementation DDBufferedProxy (StreamManipulation)
+    if ([self->_fileLogger isOnInternalLoggerQueue]) {
+        block();
+    } else {
+        dispatch_sync(self->_fileLogger.loggerQueue, block);
+    }
+}
+
+#pragma mark - Buffering
 
 - (void)flushBuffer {
-    [_bufferStream close];
-    _bufferStream = nil;
-    _bufferSize = 0;
+    [_buffer close];
+    _buffer = [NSOutputStream outputStreamToMemory];
+    [_buffer open];
+    _currentBufferSizeBytes = 0;
 }
 
-- (void)dumpBufferToDisk {
-    NSData *data = [_bufferStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
-    [self.instance logData:data];
+- (void)lt_sendBufferedDataToFileLogger {
+    NSData *data = [_buffer propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
+    [_fileLogger lt_logData:data];
     [self flushBuffer];
 }
 
-- (void)appendToBuffer:(NSData *)data {
-    __auto_type length = data.length;
-    if (data.length != 0) {
-        if (_bufferStream == nil) {
-            _bufferStream = [[NSOutputStream alloc] initToMemory];
-            [_bufferStream open];
-            _bufferSize = 0;
-        }
-        const uint8_t *appendedData = calloc(length, sizeof(uint8_t));
-        if (appendedData != NULL) {
-            [data getBytes:(void *)appendedData length:length];
-            [_bufferStream write:appendedData maxLength:length];
-            _bufferSize += length;
+#pragma mark - Logging
 
-            free((void *)appendedData);
-        }
+- (void)logMessage:(DDLogMessage *)logMessage {
+    NSData *data = [_fileLogger lt_dataForMessage:logMessage];
+    NSUInteger length = data.length;
+    if (length == 0) {
+        return;
     }
-}
 
-- (BOOL)isBufferFull {
-    return _bufferSize > self.maximumBytesCountInBuffer;
+#ifdef DEBUG
```

Comment:
This could also be:
```
#ifndef DEBUG
__unused
#endif
NSInteger written = [_buffer write:[data bytes] maxLength:length];
NSAssert(written == (NSInteger)length, @"Failed to write to memory buffer.");
```

Reply:
Nice!